### PR TITLE
Adding support for permission boundaries for AWS IAM Roles

### PR DESCRIPTION
**Pull request by victorfrancax1**
Refs #7286 
Also adding Terrraform and Cloudformation support.

### CODE REVIEW

**Comment by rifelpet** [comment]
overall this looks good. Can you add an integration test case for this? It should just involve setting the new api field with a dummy value in the two `in-*.yaml` files [here](https://github.com/kubernetes/kops/tree/master/tests/integration/update_cluster/complex) and running `./hack/update-expected.sh` and the terraform and cloudformation files in the same directory should be updated to match. 

My only question is around the API design. Would it make sense to ever have separate permission boundaries for the master role vs the node role? If so we'd want the API to be able to support that.

**Comment by victorfrancax1** [comment]
Hey @rifelpet , thanks for reviewing! I added an integrated test case as requested.

About the API design: although being possible, it's seems very unlikely to me that people would need to set permission boundaries per instance role, since they were designed mainly to prevent privilege escalation. The furthest I would go personally is configuring different boundaries for each cluster, which is supported by this implementation. As stated by the author of the original issue (can't tag him here for some reason), should an edge case arise, we could suggest that they provide their own roles to `kops`. What do you think?

File: upup/pkg/fi/cloudup/awstasks/iamrole.go [hunk 1]
```diff
@@ -147,6 +148,10 @@ func (_ *IAMRole) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *IAMRole) error
 		request.AssumeRolePolicyDocument = aws.String(policy)
 		request.RoleName = e.Name
 
+		if e.PermissionsBoundary != nil {
+			request.PermissionsBoundary = e.PermissionsBoundary
```
Comment (rifelpet) [inline upup/pkg/fi/cloudup/awstasks/iamrole.go:152]: Up in the Find function above (~ line 74) you'll need to read the PermissionsBoundary field from the AWS SDK's Role and set it on the `awstasks.IAMRole`, similar to name and id fields on those lines.

Comment (victorfrancax1) [reply upup/pkg/fi/cloudup/awstasks/iamrole.go:155]: Fixed. Thanks for pointing that out :)

Comment (victorfrancax1) [reply upup/pkg/fi/cloudup/awstasks/iamrole.go:155]: I realized that I wasn't addressing possible PermissionsBoundary updates in the `RenderAws` function. Added that logic too.

File: pkg/model/iam.go [hunk 2]
```diff
@@ -110,6 +110,11 @@ func (b *IAMModelBuilder) buildIAMTasks(igRole kops.InstanceGroupRole, iamName s
 				RolePolicyDocument: fi.WrapResource(rolePolicy),
 				ExportWithID:       s(strings.ToLower(string(igRole)) + "s"),
 			}
+
+			if b.Cluster.Spec.IAM.PermissionsBoundary != nil {
```
Comment (rifelpet) [inline pkg/model/iam.go:114]: I think we'll need an extra nil check here
```suggestion
			if b.Cluster.Spec.IAM != nil && b.Cluster.Spec.IAM.PermissionsBoundary != nil {
```

**Comment by rifelpet** [comment]
one last comment and I think this will be good to merge :) thanks for adding this!

File: pkg/model/iam.go (hunk 2 above)
Comment (victorfrancax1) [reply pkg/model/iam.go]: Fixed :)

**Comment by rifelpet** [comment]
/lgtm
/approve

This will land in Kops 1.19. I believe there will be another alpha release soon if you want to use it in an official build.

Thanks again!

**Comment by rifelpet** [comment]
One more thing @victorfrancax1 would you mind adding some docs [here](https://github.com/kubernetes/kops/blob/master/docs/iam_roles.md)? This can be done in a separate PR.

**Pull request merged**